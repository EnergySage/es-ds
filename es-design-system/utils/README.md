# /utils/ directory

This directory contains javascript utility functions and methods used directly by Vue/Nuxt
Please do not create `index.js`, group new methods and functions to the correct namespace.
If a "correct" namespace does not exist create a new file.
