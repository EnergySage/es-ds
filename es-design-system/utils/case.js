// eslint-disable-next-line import/prefer-default-export
export const kebabCase = (str) => str
    .match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
    .join('-')
    .toLowerCase();
