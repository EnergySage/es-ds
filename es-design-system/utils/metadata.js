// import sassThemeColors from '@energysage/es-bs-extends/scss/variables/_theme-colors.scss';

// Possible Component Sizes
export const designStages = ['Atom', 'Molecule', 'Organism', 'Template', 'Page'];

// TODO: font-family values should come from: '@energysage/es-bs-extends/scss/variables/_font-family.scss';
const fontFamilies = {
    // eslint-disable-next-line max-len
    fontFamilySansSerif: '"Inter", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", "Liberation Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"',
    fontFamilyMonospace: 'SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace',
};
fontFamilies.fontFamilyBase = fontFamilies.fontFamilySansSerif;

// const themeColors = sassThemeColors;

export const bsVariables = {
    ...fontFamilies,
    // themeColors,
};

export default {
    designStages,
    bsVariables,
};
