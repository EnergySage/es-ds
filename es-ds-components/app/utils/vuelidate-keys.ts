/**
 * constants.js
 *
 * String constants that can be used when defining validations objects
 */

const REQUIRED = 'vuelidateRequired';
const MIN_VALUE = 'vuelidateMinValue';
const MAX_VALUE = 'vuelidateMaxValue';
const REQUIRED_IF = 'vuelidateRequiredIf';
const NUMERIC = 'vuelidateNumeric';
const INTEGER = 'vuelidateInteger';
const MIN_LENGTH = 'vuelidateMinLength';
const MAX_LENGTH = 'vuelidateMaxLength';
const HAS_NUMBER = 'vuelidateHasNumber';
const HAS_SPECIAL_CHARACTER = 'vuelidateHasSpecialCharacter';
const HAS_UPPERCASE_LETTER = 'vuelidateHasUppercaseLetter';
const HAS_LOWERCASE_LETTER = 'vuelidateHasLowercaseLetter';
const EMAIL = 'vuelidateEmail';
const PHONE = 'vuelidatePhone';
const HELPERS = 'vuelidateHelpers';

export default {
    REQUIRED,
    REQUIRED_IF,
    MIN_VALUE,
    MAX_VALUE,
    MIN_LENGTH,
    MAX_LENGTH,
    NUMERIC,
    INTEGER,
    EMAIL,
    PHONE,
    HAS_NUMBER,
    HAS_SPECIAL_CHARACTER,
    HAS_UPPERCASE_LETTER,
    HAS_LOWERCASE_LETTER,
    HELPERS,
};
