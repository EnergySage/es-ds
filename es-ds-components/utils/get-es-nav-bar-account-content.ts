// rooftop account links
export default (
    ES_DOMAIN = process.env.ES_DOMAIN || 'https://www.energysage.com',
    ACCOUNTS_DOMAIN = process.env.ACCOUNTS_DOMAIN || 'https://accounts.energysage.com',
    REGISTER_LINK = process.env.REGISTER_LINK || `${ACCOUNTS_DOMAIN}/register/?next=${ES_DOMAIN}/register-thanks/`,
) => ({
    mobileAccountButtonAltText: 'Open account menu',
    loggedIn: {
        items: [
            {
                name: 'My profile',
                link: `${ES_DOMAIN}/profile/`,
            },
            {
                name: 'My quotes',
                link: `${ES_DOMAIN}/your-marketplace/`,
            },
            {
                name: 'My messages',
                link: `${ES_DOMAIN}/message/`,
            },
            {
                name: 'My project preferences',
                link: `${ES_DOMAIN}/market/solarpv/update/`,
            },
            {
                name: 'Account settings',
                link: `${ES_DOMAIN}/account-settings/router/`,
            },
            {
                name: 'How it works',
                link: `${ES_DOMAIN}/market-process/`,
            },
            {
                name: 'Write a review',
                link: `${ES_DOMAIN}/share-your-experience/`,
            },
            {
                name: 'Sign out',
                link: `${ES_DOMAIN}/logout/`,
            },
        ],
    },
    loggedOut: {
        menuButtonText: 'Sign in',
        signIn: {
            name: 'Sign in',
            link: `${ES_DOMAIN}/login/`,
        },
        createAccount: {
            name: 'Create an account',
            link: REGISTER_LINK,
        },
    },
});
