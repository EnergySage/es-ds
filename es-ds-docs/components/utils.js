// utils/loadFile.js
export async function loadFile(filePath) {
    try {
      const module = await import(`${filePath}.vue?raw`);
      return module;
    } catch (error) {
      console.error('Error loading file:', error);
      return null;
    }
  }
