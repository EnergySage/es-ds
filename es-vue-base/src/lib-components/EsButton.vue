<template>
    <b-button
        :class="{ 'inline': inline }"
        :variant="computedVariant"
        :size="size"
        v-bind="$attrs"
        v-on="$listeners">
        <!--
            @slot Button Content
            @binding {string} text or html of the button content
        -->
        <slot />
    </b-button>
</template>

<script lang="js">
import { BButton } from 'bootstrap-vue';

export default {
    name: 'EsButton',
    components: {
        BButton,
    },
    props: {
        /**
         * Button Variant
         *
         */
        variant: {
            type: String,
            default: 'primary',
            required: false,
        },
        /**
         * Outline only button
         */
        outline: {
            type: Boolean,
            default: false,
        },
        /**
         * Size of button
         */
        size: {
            type: String,
            validator: (val) => ['lg', 'md', 'sm'].includes(val),
            default: 'md',
        },
        /**
         * Only works for 'link' variant buttons.
         * Removes the fixed height and padding so the button can fit nicely within a block of text.
         */
        inline: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        /**
         * Compute the Variant
         * used for when you want to `outline` something but don't know the
         * theme's default variant
         *
         * @returns { string } computed variant based on outline prop
         */
        computedVariant() {
            return `${this.outline && this.variant !== 'link' ? 'outline-' : ''}${this.variant}`;
        },
    },
};
</script>

<style lang="scss" scoped>
.btn-link.inline {
    /* use normal CSS here so users can override with utility classes as necessary */
    border: 0;
    height: auto;
    line-height: inherit;
    margin: 0;
    padding: 0;
    vertical-align: baseline;
}
</style>
