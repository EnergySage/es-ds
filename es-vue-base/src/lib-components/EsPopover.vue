<template>
    <b-popover
        ref="popover"
        :show="show"
        :target="target"
        :placement="placement"
        :triggers="triggers"
        :custom-class="`es-popover-${variant}`">
        <template
            v-if="hasTitle"
            #title>
            <!--
            @slot Title
            @binding {string} text or html of the title content
            -->
            <slot name="title" />
            <es-button
                inline
                variant="link"
                class="es-popover-close p-0 float-right"
                @click="onClose">
                <XIcon
                    height="20px"
                    width="20px" />
            </es-button>
        </template>
        <!--
        @slot Popover Content
        @binding {string} text or html of the popover content
        -->
        <template #default>
            <div class="d-flex">
                <slot />
                <es-button
                    v-if="!hasTitle"
                    inline
                    variant="link"
                    class="es-popover-close p-0 pl-50"
                    @click="onClose">
                    <XIcon
                        height="20px"
                        width="20px" />
                </es-button>
            </div>
        </template>
    </b-popover>
</template>

<script lang="js">
import { BPopover } from 'bootstrap-vue';
import XIcon from '../lib-icons/icon-x.vue';
import EsButton from './EsButton.vue';

export default {
    name: 'EsPopover',
    components: {
        BPopover,
        EsButton,
        XIcon,
    },
    props: {
        /**
         * Target
         * (required)
         */
        target: {
            type: String,
            required: true,
        },
        /**
         * Triggers
         * (optional, defaults to 'focus')
         */
        triggers: {
            type: [String, Object],
            required: false,
            default: 'focus',
        },
        /**
         * Placement
         * (optional, defaults to 'auto')
         */
        placement: {
            type: String,
            required: false,
            // eslint-disable-next-line max-len
            validator: (val) => ['auto', 'top', 'bottom', 'right', 'left', 'topleft', 'topright', 'bottomleft', 'bottomright', 'lefttop', 'leftbottom', 'righttop', 'rightbottom'].includes(val),
            default: 'auto',
        },
        /**
         * Show
         * (optional, defaults to 'false')
         */
        show: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * Variant
         */
        variant: {
            type: String,
            required: false,
            default: 'dark',
            validator: (val) => ['light', 'dark'].includes(val),
        },
    },
    computed: {
        hasTitle() {
            return !!this.$slots.title;
        },
    },
    methods: {
        onClose() {
            this.$refs.popover.$emit('close');
        },
    },
};
</script>

<style lang="scss">
@use "~@energysage/es-bs-base/scss/variables" as variables;

/* stylelint-disable no-descending-specificity */
.es-popover-light {
    &.popover {
        background-color: variables.$white;
        border: 1px solid variables.$white;
    }
    // styling for all arrow backgrounds
    &.bs-popover-bottom, &.bs-popover-auto[x-placement^=bottom] {
        > .arrow {
            &::before {
                border-bottom-color: variables.$white;
            }

            &::after {
                border-bottom-color: variables.$white;
                top: 1px;
            }
        }
    }

    &.bs-popover-top, &.bs-popover-auto[x-placement^=top] {
        > .arrow {
            &::before {
                border-top-color: variables.$white;
            }

            &::after {
                border-top-color: variables.$white;
                bottom: 1px;
            }
        }
    }

    &.bs-popover-right, &.bs-popover-auto[x-placement^=right] {
        > .arrow {
            &::before {
                border-right-color: variables.$white;
            }

            &::after {
                border-right-color: variables.$white;
                left: 1px;
            }
        }
    }

    &.bs-popover-left, &.bs-popover-auto[x-placement^=left] {
        > .arrow {
            &::before {
                border-left-color: variables.$white;
            }

            &::after {
                border-left-color: variables.$white;
                right: 1px;
            }
        }
    }

    .popover-header {
        background-color: variables.$white;
        border-bottom-color: variables.$white;
        color: variables.$gray-900;

        &::before {
            border-bottom-color: variables.$white;
        }
    }

    .popover-body {
        color: variables.$gray-900;
    }

    .es-popover-close {
        color: variables.$gray-900;
    }
}

.es-popover-dark {
    .es-popover-close, a {
        color: variables.$white;
    }
}
</style>
