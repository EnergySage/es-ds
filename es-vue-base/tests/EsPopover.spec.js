/* eslint-disable no-console */
import { mount } from '@vue/test-utils';
import EsPopover from '@/src/lib-components/EsPopover.vue';
import EsButton from '@/src/lib-components/EsButton.vue';
import XIcon from '@/src/lib-icons/x.vue';
import { waitRAF, waitNT } from './utils';

// https://github.com/bootstrap-vue/bootstrap-vue/blob/dev/src/components/popover/popover.spec.js

// Our test application definition
const App = {
    props: [
        'triggers',
        'show',
        'target',
        'customClass',
        'titleAttr',
        'btnDisabled',
    ],
    render(h) {
        return h('article', { attrs: { id: 'wrapper' } }, [
            h(
                'button',
                {
                    attrs: {
                        id: 'triggerButton',
                        type: 'button',
                        disabled: this.btnDisabled || null,
                        title: this.titleAttr || null,
                    },
                },
                'text',
            ),
            h(
                EsPopover,
                {
                    attrs: {
                        id: 'esPopover',
                    },
                    props: {
                        target: 'triggerButton',
                        triggers: this.triggers,
                        show: this.show,
                        customClass: this.customClass,
                    },
                },
                [h('template', { slot: 'title' }, this.$slots.title), this.$slots.default || ''],
            ),
        ]);
    },
};

describe('EsPopover', () => {
    const originalCreateRange = document.createRange;
    const origGetBCR = Element.prototype.getBoundingClientRect;

    beforeEach(() => {
        // https://github.com/FezVrasta/popper.js/issues/478#issuecomment-407422016
        // Hack to make Popover not bork out during tests
        // Note popover still does not do any positioning calculation in JSDOM though
        // So we cannot test actual positioning, just detect when it is open
        document.createRange = () => ({
            setStart: () => {},
            setEnd: () => {},
            commonAncestorContainer: {
                nodeName: 'BODY',
                ownerDocument: document,
            },
        });
        // Mock `getBoundingClientRect()` so that the `isVisible(el)` test returns `true`
        // Needed for visibility checks of trigger element, etc
        Element.prototype.getBoundingClientRect = jest.fn(() => ({
            width: 24,
            height: 24,
            top: 0,
            left: 0,
            bottom: 0,
            right: 0,
        }));
    });

    afterEach(async () => {
        // Reset overrides
        jest.setTimeout(() => {
            document.createRange = originalCreateRange;
            Element.prototype.getBoundingClientRect = origGetBCR;
            waitRAF();
        }, 10000);
    });

    it('has expected default structure and props', async () => {
        const wrapper = mount(App, {
            attachTo: document.body,
            propsData: {
                triggers: 'click',
                customClass: 'es-popover-light',
            },
            slots: {
                title: 'title',
                default: 'content',
            },
        });
        expect(wrapper.vm).toBeDefined();
        await waitNT(wrapper.vm);

        expect(wrapper.element.tagName).toBe('ARTICLE');
        expect(wrapper.attributes('id')).toBeDefined();
        expect(wrapper.attributes('id')).toBe('wrapper');

        // The trigger button
        const $button = wrapper.find('button');
        expect($button.exists()).toBe(true);
        expect($button.attributes('id')).toBeDefined();
        expect($button.attributes('id')).toBe('triggerButton');
        expect($button.attributes('aria-describedby')).toBeUndefined();

        // <es-popover> wrapper
        const $popover = wrapper.findComponent(EsPopover);
        expect($popover.exists()).toBe(true);
        expect($popover.element.nodeType).toEqual(Node.COMMENT_NODE);
        expect($popover.selector.props.show.default).toBe(false);
        expect($popover.selector.props.placement.default).toBe('auto');
        expect($popover.selector.props.triggers.default).toBe('focus');
        expect($popover.selector.props.customClass.default).toBe('es-popover-dark');
        expect($popover.props().target).toBe('triggerButton');
        expect($popover.props().customClass).toBe('es-popover-light');

        // Close button on popover
        const $closeButton = $popover.findComponent(EsButton);
        expect($closeButton).not.toBeNull();

        // XIcon on popover
        const $xIcon = $popover.findComponent(XIcon);
        expect($xIcon).not.toBeNull();

        wrapper.destroy();
    });

    it('initially open has expected structure', async () => {
        jest.useFakeTimers();
        const wrapper = mount(App, {
            attachTo: document.body,
            propsData: {
                triggers: 'click',
                show: true,
            },
            slots: {
                title: 'title',
                default: 'content',
            },
        });

        expect(wrapper.vm).toBeDefined();
        jest.setTimeout(() => {
            waitNT(wrapper.vm);
            waitRAF();
            waitNT(wrapper.vm);
            waitRAF();
            jest.runOnlyPendingTimers();
            waitNT(wrapper.vm);
        }, 10000);

        expect(wrapper.element.tagName).toBe('ARTICLE');
        expect(wrapper.attributes('id')).toBeDefined();
        expect(wrapper.attributes('id')).toBe('wrapper');

        // The trigger button
        const $button = wrapper.find('button');
        expect($button.exists()).toBe(true);
        expect($button.attributes('id')).toBeDefined();
        expect($button.attributes('id')).toBe('triggerButton');

        // <es-popover> wrapper
        const $popover = wrapper.findComponent(EsPopover);
        expect($popover.props().show).toBe(true);
        expect($popover.exists()).toBe(true);
        expect($popover.element.nodeType).toEqual(Node.COMMENT_NODE);
        expect($popover.props().show).toBe(true);

        // Hide the Popover
        await wrapper.setProps({
            show: false,
        });
        jest.setTimeout(() => {
            waitNT(wrapper.vm);
            waitRAF();
            waitNT(wrapper.vm);
            waitRAF();
            jest.runOnlyPendingTimers();
        }, 10000);

        // Popover element should not be in the document
        expect($popover.props().show).toBe(false);

        wrapper.destroy();
    });
});
