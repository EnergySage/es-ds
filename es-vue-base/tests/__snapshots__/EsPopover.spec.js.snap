// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsPopover <EsPopover /> 1`] = `"<span target=\\"target\\" show=\\"true\\"><button id=\\"myTarget\\" type=\\"button\\" class=\\"btn btn-primary btn-md\\">Button</button><!----></span>"`;

exports[`EsPopover Close button exists 1`] = `"<span target=\\"target\\"><button id=\\"myTarget\\" type=\\"button\\" class=\\"btn btn-primary btn-md\\">Button</button><!----></span>"`;

exports[`EsPopover Content slot exists 1`] = `"<span target=\\"target\\"><button id=\\"myTarget\\" type=\\"button\\" class=\\"btn btn-primary btn-md\\">Button</button><!----></span>"`;

exports[`EsPopover Title slot exists 1`] = `"<span target=\\"target\\"><button id=\\"myTarget\\" type=\\"button\\" class=\\"btn btn-primary btn-md\\">Button</button><!----></span>"`;
